### PR TITLE
Contracts:  update wasmi to 0.32

### PR DESCRIPTION
**Pull request by pgherveou**
take over #2941 
[Weights compare](https://weights.tasty.limo/compare?unit=weight&ignore_errors=true&threshold=10&method=asymptotic&repo=polkadot-sdk&old=master&new=pg%2Fwasmi-to-v0.32.0-beta.7&path_pattern=substrate%2Fframe%2F**%2Fsrc%2Fweights.rs%2Cpolkadot%2Fruntime%2F*%2Fsrc%2Fweights%2F**%2F*.rs%2Cpolkadot%2Fbridges%2Fmodules%2F*%2Fsrc%2Fweights.rs%2Ccumulus%2F**%2Fweights%2F*.rs%2Ccumulus%2F**%2Fweights%2Fxcm%2F*.rs%2Ccumulus%2F**%2Fsrc%2Fweights.rs)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5517795

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5517830

**Comment by pgherveou** [comment]
bot help

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by mordamax** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by deuszx** [comment]
Hey @pgherveou , I saw that the benchmarks have finished but I'm not sure how to interpret the results. Are there any findings regarding possible execution improvements here?

**Comment by pgherveou** [comment]
@Robbepop @athei not sure why we commented on the previous PR
https://github.com/paritytech/polkadot-sdk/pull/2941#pullrequestreview-1836267022

> We need to change the benchmarks to compile the code in its setup so that when executing the contract it is only execution.

Why would we want to do that when we benchmark the extrinsic that is suppose to measure the whole thing

File: substrate/frame/contracts/src/wasm/mod.rs [hunk 1]
```diff
@@ -201,19 +206,14 @@ impl<T: Config> WasmBlob<T> {
 	/// When validating we pass `()` as `host_state`. Please note that such a dummy instance must
 	/// **never** be called/executed, since it will panic the executor.
 	pub fn instantiate<E, H>(
```
Comment (Robbepop) [inline substrate/frame/contracts/src/wasm/mod.rs:208]: For enabling skipping of Wasm validation what I would have expected is to add another parameter to this function, e.g. `validate: bool`. And based on `validate` you'd either call `LoadedModule::new` or `LoadedModule::new_unchecked`. For `upload_code` we'd set `validate` to `true` and for execution we'd set `validate` to `false`. This should require the minimum amount of changes needed to introduce this feature.

**Review by Robbepop — CHANGES_REQUESTED** [review]

File: substrate/frame/contracts/src/wasm/mod.rs (hunk 1 above)
Comment (pgherveou) [reply substrate/frame/contracts/src/wasm/mod.rs:208]: I splitted the responsability, between validate and instantiate.
This also helped spotting a potential bug where we were loading (unnecessarily) twice the module for validation 

so tldr we build and pass an unchecked module when we execute the code, we pass a checked module when we just want to validate

Comment (pgherveou) [reply substrate/frame/contracts/src/wasm/mod.rs:208]: I will cherry pick these changes on a clean PR and rebase this one on top so we can review that separately

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5612759

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5613061

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5613648

**Comment by pgherveou** [comment]
bot cancel 1-220b369f-437d-46dd-b2ab-c07221359aaa

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5614105

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5616461

**Comment by deuszx** [comment]
Not sure if I'm reading the [new gas benchmarks](https://weights.tasty.limo/compare?unit=weight&ignore_errors=true&threshold=10&method=asymptotic&repo=polkadot-sdk&old=master&new=pg%2Fwasmi-to-v0.32.0-beta.7&path_pattern=substrate%2Fframe%2F**%2Fsrc%2Fweights.rs%2Cpolkadot%2Fruntime%2F*%2Fsrc%2Fweights%2F**%2F*.rs%2Cpolkadot%2Fbridges%2Fmodules%2F*%2Fsrc%2Fweights.rs%2Ccumulus%2F**%2Fweights%2F*.rs%2Ccumulus%2F**%2Fweights%2Fxcm%2F*.rs%2Ccumulus%2F**%2Fsrc%2Fweights.rs) correct but it seems that quite a few methods have regressed >10%.

File: substrate/frame/contracts/src/wasm/prepare.rs [hunk 2]
```diff
@@ -263,17 +268,25 @@ where
...
+						LoadingMode::Checked,
+						CompilationMode::Eager,
+					)?
 				},
 			Determinism::Enforced => LoadedModule::new::<T>(
 				code,
 				Determinism::Enforced,
 				stack_limits,
 				LoadingMode::Checked,
+				CompilationMode::Eager,
```
Comment (athei) [inline substrate/frame/contracts/src/wasm/prepare.rs:289]: What is the reasoning to compile eagerly when validating? Is it possible for lazy compilation to fail on a validated (but not fully compiled) module? @Robbepop

**Review by athei — COMMENTED** [review]
Looks good. We will need to re-work the `WasmModule::<T>::sized()` benchmark. probably.

We should change it to be the worst case for lazy compilation and check how bad it actually is. The worst case would be to generate only empty functions.

File: substrate/frame/contracts/src/wasm/prepare.rs (hunk 2 above)
Comment (Robbepop) [reply substrate/frame/contracts/src/wasm/prepare.rs:289]: When using lazy compilation we also defer Wasmi compilation errors e.g. when exceeding specific Wasmi compilation limits.

Comment (athei) [reply substrate/frame/contracts/src/wasm/prepare.rs:289]: Okay then we have to check eagerly.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5866028

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5868367

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5873478

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot cancel

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot cancel

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by athei** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by athei** [comment]
Looks good except for the `seal_instantiate` benchmark. Base weight goes up massively and the `t` parameter is no longer in use.

Screenshot 2024-05-24 at 13 03 30

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bench-all pallet --pallet=pallet_contracts

**Comment by pgherveou** [comment]
bot bench-all pallet --pallet=pallet_contracts

File: prdoc/pr_3679.prdoc [hunk 3]
```diff
@@ -0,0 +1,12 @@
+title: "[pallet-contracts] bump wasmi to 0.32"
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      - Bump wasmi to 0.32
+      - Turn on lazy and unchecked compilation when calling a contract.
+      See https://docs.rs/wasmi/0.32.0-beta.17/wasmi/enum.CompilationMode.html#variant.Lazy
+      See https://docs.rs/wasmi/0.32.0-beta.17/wasmi/struct.Module.html#method.new_unchecked
```
Comment (Robbepop) [inline prdoc/pr_3679.prdoc:9]: needs to be updated to stable `0.32.0`

**Comment by pgherveou** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Review by athei — APPROVED** [review]

**Pull request merged**